### ISSUE TITLE
Add "print message" Msg to RE

### ISSUE THREAD

**Issue by tacaswell** [bluesky/bluesky#1379]
As we move to a our mostly-remote collection, something that has come up with our BL staff is the ability to publish text messages to _something_ as a way to get user-controlled text out to the world (e.g. slack).   While we can extract a lot of information the documents to re-publish, but having the ability to publish arbitrary non-data messages out of the RE.  This could be generic status messages generated by plans or specific messages that user plans may want to publish.   I propose adding an additional message and callback registry to the RE for user-control message emission.

We do want to this as a message so that this will be safe to run in list / preview / pre-processors etc. 

I propose the following message layout:

```python
Msg('notice', None, (payload, ), {"con': "text, "dict": "ionary"}, run)
```

and then the registry would take signatures of the form

```python

async def handle_notice(payload, context):
    await post(payload)

```

You could then wrap slack / teams / generic post request.  

We would need to add

```python

async def _notice(self, msg):
    payload, = msg.arg
    context = msg.kwargs
    # push this to a different task?
    return asncyio.create_task(self._notice_callback_registiry.emit(payload, context))

def subscribe_notice(self, callback):
    return self._notice_callback_registry.subscribe(callback)

def unsubscribe_notice(self, cid):
    return self._notice_callback.unsubscribe(cid)

```

to the RunEnigne  and a class that looks like

```python
class NoticeRegistry:
    def __init__(self):
        self._counter = itertools.count()
        self._callbacks = {}
    def subscribe(self, callback):
        cid = next(self._counter)
        self._callbacks[cid] = callback   # modulo wrapping / weakref
    def unsubscribe(self, cid):
        self._callbacks.pop(cid, None)
    def emit(self, payload, context):
        for cb in self._callbacks.values():
            try:
                await cb(payload, context)
            except Exception:
                 print("womp womp, what should we do here?")
    


```

Open questions:

 - in the signature should it be `**context` or `context` (that is we expect the signature to to varkwargs or the context dictionary a a second positional arguement)?  I think that it is better to be a positional argument because taking a dictionary that may have excess keys seems better than having to ignore kwargs.  Additionally, if we did `**` the context dictionary in we could end up with registered functions that require a give key in the signature which would fail in the _calling_ code (rather than is the user function).  While you can still write a function that requires a key in the dictionary at least the failure will happen in the user code and they can deal with it (or not) as they choose).
 - should the signature be async? (I think yes so we don't block on network I/O)
 - name? Other options {"notice", "log", "post", "message"}

------

Alternate proposal:

Add a method to add things to a particular logger on the RE

```python
Msg('log', None, (payload, level), context_dict, run))
```

and the corresponding method would be

```python
async def _log(self, msg):
    payload, level = msg.args
    self.logger.log(level, payload, extra=msg.kwargs)
``` 

and then rely on the logging framework to do the dispatch. The pro-side is that we get a lot of stuff for free, the down side is we get a whole bunch of complexity.  With this proposal I think there are questions:
 - should we allow the user to pass in what logger to send to?
 - can we get this async again?

---

I suspect we may actually want _both_ of these options.

**Comment by danielballan** [comment]
I agree we may eventually having both, but I propose we start with the logger-based thing and wait until we really need something else before adding another registry.
